### PR TITLE
Use SecretKeySelector for AM configuration

### PR DESCRIPTION
**Pull request by pintohutch**
- Use SecretKeySelector over string to load Alertmanager configuration from secret.
- Add some unit tests around defaulting behavior.
- nits: changed some default names.

### CODE REVIEW

File: cmd/operator/deploy/crds/monitoring.googleapis.com_operatorconfigs.yaml [hunk 1]
```diff
@@ -105,15 +105,31 @@ spec:
...
                   file.
-                type: string
+                properties:
+                  key:
+                    description: The key of the secret to select from.  Must be a
+                      valid secret key.
+                    type: string
+                  name:
+                    description: 'Name of the referent. More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
+                      TODO: Add other useful fields. apiVersion, kind, uid?'
```
Comment (realschwa) [inline cmd/operator/deploy/crds/monitoring.googleapis.com_operatorconfigs.yaml:125]: should this be a comment?

Comment (pintohutch) [reply cmd/operator/deploy/crds/monitoring.googleapis.com_operatorconfigs.yaml:125]: This is actually [pulled in](https://github.com/GoogleCloudPlatform/prometheus-engine/blob/24a44d0cf189f37126bda692c4c74e272ff24b2e/vendor/k8s.io/api/core/v1/types.go#L5609) from the vendored k8s api go library.

File: cmd/operator/deploy/crds/monitoring.googleapis.com_operatorconfigs.yaml [hunk 2]
```diff
@@ -105,15 +105,31 @@ spec:
...
+                    type: string
+                  optional:
+                    description: Specify whether the Secret or its key must be defined
+                    type: boolean
+                required:
```
[the remark this answers is not in the document]
Comment (realschwa) [reply cmd/operator/deploy/crds/monitoring.googleapis.com_operatorconfigs.yaml:130]: I think this is also pulled in from https://github.com/GoogleCloudPlatform/prometheus-engine/blob/24a44d0cf189f37126bda692c4c74e272ff24b2e/vendor/k8s.io/api/core/v1/types.go#L2058 ?

[the remark this answers is not in the document]
Comment (pintohutch) [reply cmd/operator/deploy/crds/monitoring.googleapis.com_operatorconfigs.yaml:130]: Interesting. For some reason, the `LocalObjectReference` API (which is embedded in the `SecretKeySelector`) [explicitly calls out](https://github.com/GoogleCloudPlatform/prometheus-engine/blob/24a44d0cf189f37126bda692c4c74e272ff24b2e/vendor/k8s.io/api/core/v1/types.go#L5610) name as optional?

You can't create a resource without a name in k8s, so I'm not sure what that's about. The `name` field can actually be left blank, in which case the operator logs:
```
alertmanager config secret not found in namespace gmp-public: unable to get secret \"\": Secret \"\" not found
```
which I think is sufficient signal for now. Note - this issue applies in everywhere we're using the `SecretKeySelector` already.

We should add some validating webhooks to ensure this field is not omitted to surface this better to the user. I think that is a feature we can add later. Thanks for calling this out!

[the remark this answers is not in the document]
Comment (xichen2020) [reply cmd/operator/deploy/crds/monitoring.googleapis.com_operatorconfigs.yaml:130]: Interesting, thanks for the analysis. SGTM.

**Review by xichen2020 — APPROVED** [review]

**Review by realschwa — APPROVED** [review]
lgtm

**Pull request merged**